README.md
# pull_refresh_widget:


####  A Flutter package allows you to add pull refresh widget as listview builder or column 
[![pub package](https://img.shields.io/pub/v/arabic_font.svg)](https://pub.dev/packages/pull_refresh_widget)   

------------------
## # 👨‍💻 *Developed  by:*

<img alt="profile" src="https://abom.me/packages/profile.png" width="50" height="50"  style=" border-radius: 100%"/>

**Nasr Al-Rahbi [@abom_me](https://twitter.com/abom_me)**

## 👨🏻‍💻 Find me in  :
[![Twitter](https://img.shields.io/badge/Twitter-%231DA1F2.svg?logo=Twitter&logoColor=white)](https://twitter.com/abom_me)
[![Instagram](https://img.shields.io/badge/Instagram-%23E4405F.svg?logo=Instagram&logoColor=white)](https://instagram.com/abom.me)
[![LinkedIn](https://img.shields.io/badge/LinkedIn-%230077B5.svg?logo=linkedin&logoColor=white)](https://linkedin.com/in/nasr-al-rahbi-08a573245)
[![Stack Overflow](https://img.shields.io/badge/-Stackoverflow-FE7A16?logo=stack-overflow&logoColor=white)](https://stackoverflow.com/users/19994059/nasr-al-rahbi)

---------------
<br>



|             | Android | iOS | Flutter Web |
| ----------- |---------|-----|-------------|
| **Support** | Yes     | Yes | Yes         |




---
## Preview:

<img alt="changing fonts with arabic_font and hot reload" src="https://abom.me/packages/pull_refresh.gif" width="20%" />


---
## Setup

Pubspec changes:

```
      dependencies:
        pull_refresh_widget: <last_version>
```
---
### Getting Started:

Sample example:

```dart
import 'package:pull_refresh_widget/pull_refresh_widget.dart';
ScrollController _controller=ScrollController();
String text="Pull To Refresh";
RefreshScrollColumn(


controller: _controller,
onRefresh: () {
print('Refreshing');

},
onRefreshEnd: (){
print('Refreshed');
setState(() {
text='Refreshed';
});
},
children: [
Container(
alignment: Alignment.center,
height: 100,
width: 300,
child:  Text(text,style: TextStyle(color: Colors.red,fontSize: 30),),
)
],

),
```
---
- What You can Do:
    - use RefreshScrollColumn() to add it as Column
    - use RefreshListViewBuilder() to add it as ListView.builder


---


### To use RefreshScrollColumn():

```dart

import 'package:pull_refresh_widget/pull_refresh_widget.dart';
ScrollController _controller=ScrollController();
String text="Pull To Refresh";
RefreshScrollColumn(


controller: _controller,
onRefresh: () {
print('Refreshing');

},
onRefreshEnd: (){
print('Refreshed');
setState(() {
text='Refreshed';
});
},
children: [
Container(
alignment: Alignment.center,
height: 100,
width: 300,
child:  Text(text,style: TextStyle(color: Colors.red,fontSize: 30),),
)
],

),
```

---

<br>

### To use RefreshListViewBuilder() :

```dart
RefreshListViewBuilder(

controller: _controller,
onRefresh: (){
print("refreshed");
}, itemBuilder: (context,index){
return Text('$index');
})
```





---


### Parameters of the RefreshScrollColumn():
````dart

/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
/// ###  Add scroll controller  to can control the scroll widget
/// 
/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
/// ### اضف 
/// ### scroll controller
///### لكي تستطيع التحكم بالويدجت

final ScrollController controller ;

/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
/// ###  Add Here Widgets you want to show in the screen
/// 
/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
/// ### هنا اضف المكونات التي تريدها بداخل الويدجت من نوع 
/// ### Widget

final  List<Widget> children;

/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
/// ###  Here the widget shown when the user pulling down
/// ###  it's  CircularProgressIndicator() by default
/// 
/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
/// ### هنا اختار الويدجت التي تريدها تظهر للمستخدم عندما يسحب للأسفل
/// ### CircularProgressIndicator() بشكل افتراضي
final Widget? loadingWidget;


/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
/// ###  Color of the arrow icons it's gray by default 
/// 
/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
/// ### هنا لون ايقونات الأسهم ، لونها رصاصي بشكل افتراضي 



final Color? iconsColor;

/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
/// ###  Add here the functions you want when the user pull down
/// 
/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
/// ### هنا اضف الأوامر التي تريدها عندما المستخدم يسحب للأسفل 

final void Function() onRefresh;

/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
/// ###  Add here the functions you want when the refreshing end
/// 
/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
/// ### هنا اضف الأوامر التي تريدها عندما ينتهي التحديث 
final void Function()? onRefreshEnd;

/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
/// ###  Here you can change the time taken to refresh before hiding the refresh widget
/// ###  it's 3 sec by default
/// 
/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
/// ### من هنا بإمكانك تغيير الوقت المستغرق في التحديث قبل لا يقوم بإخفاء ويدجت التحديث
/// ### الان يأخذ ٣ ثواني بشكل افتراضي
final Duration? refreshTime;

final MainAxisAlignment? mainAxisAlignment;
final MainAxisSize? mainAxisSize ;
final  CrossAxisAlignment? crossAxisAlignment;
final  TextDirection? textDirection;
final  VerticalDirection? verticalDirection ;
final  TextBaseline? textBaseline;

````
---
### Parameters of the RefreshListViewBuilder():
````dart


/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
/// ###  Add scroll controller  to can control the scroll widget
/// 
/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
/// ### اضف 
/// ### scroll controller
///### لكي تستطيع التحكم بالويدجت

final ScrollController controller ;



/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
/// ###  Here the widget shown when the user pulling down
/// ###  it's  CircularProgressIndicator() by default
/// 
/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
/// ### هنا اختار الويدجت التي تريدها تظهر للمستخدم عندما يسحب للأسفل
/// ### CircularProgressIndicator() بشكل افتراضي
final Widget? loadingWidget;


/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
/// ###  Color of the arrow icons it's gray by default 
/// 
/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
/// ### هنا لون ايقونات الأسهم ، لونها رصاصي بشكل افتراضي 



final Color? iconsColor;

/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
/// ###  Add here the functions you want when the user pull down
/// 
/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
/// ### هنا اضف الأوامر التي تريدها عندما المستخدم يسحب للأسفل
final void Function() onRefresh;
/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
/// ###  Add here the functions you want when the refreshing end
/// 
/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
/// ### هنا اضف الأوامر التي تريدها عندما ينتهي التحديث 
final void Function()? onRefreshEnd;

/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
/// ###  Here you can change the time taken to refresh before hiding the refresh widget
/// ###  it's 3 sec by default
/// 
/// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
/// ### من هنا بإمكانك تغيير الوقت المستغرق في التحديث قبل لا يقوم بإخفاء ويدجت التحديث
/// ### الان يأخذ ٣ ثواني بشكل افتراضي
final Duration? refreshTime;
final  NullableIndexedWidgetBuilder itemBuilder;

final  Axis scrollDirection = Axis.vertical;
final bool reverse = false;
final bool? primary;
final ScrollPhysics? physics;
final  bool shrinkWrap = false;
final  EdgeInsetsGeometry? padding;
final  double? itemExtent;
final Widget? prototypeItem;
final  int? Function(Key)? findChildIndexCallback;
final  int? itemCount;
final  bool addAutomaticKeepAlives = true;
final  bool addRepaintBoundaries = true;
final  bool addSemanticIndexes = true;
final  double? cacheExtent;
final  int? semanticChildCount;
final  DragStartBehavior dragStartBehavior = DragStartBehavior.start;
final ScrollViewKeyboardDismissBehavior keyboardDismissBehavior = ScrollViewKeyboardDismissBehavior.manual;
final String? restorationId;
final Clip clipBehavior = Clip.hardEdge;
final MainAxisAlignment? mainAxisAlignment;
final MainAxisSize? mainAxisSize ;
final  CrossAxisAlignment? crossAxisAlignment;
final  TextDirection? textDirection;
final  VerticalDirection? verticalDirection ;
final  TextBaseline? textBaseline;
````


## 👨🏻‍💻 Follow me  :
[![Twitter](https://img.shields.io/badge/Twitter-%231DA1F2.svg?logo=Twitter&logoColor=white)](https://twitter.com/abom_me)
[![Instagram](https://img.shields.io/badge/Instagram-%23E4405F.svg?logo=Instagram&logoColor=white)](https://instagram.com/abom.me)
[![LinkedIn](https://img.shields.io/badge/LinkedIn-%230077B5.svg?logo=linkedin&logoColor=white)](https://linkedin.com/in/nasr-al-rahbi-08a573245)
[![Stack Overflow](https://img.shields.io/badge/-Stackoverflow-FE7A16?logo=stack-overflow&logoColor=white)](https://stackoverflow.com/users/19994059/nasr-al-rahbi)



CHANGELOG.md
## 0.0.1

* TODO: Describe initial release.

## 0.0.2

* Fixing some bugs



lib/src/pull_refresh_column.dart
import 'dart:async';

import 'package:flutter/material.dart';

class RefreshScrollColumn extends StatefulWidget {
  const RefreshScrollColumn(
      {super.key,
      required this.controller,
      required this.onRefresh,
      required this.children,
      this.loadingWidget,
      this.iconsColor,
      this.refreshTime,
      this.textDirection,
      this.textBaseline,
      this.mainAxisAlignment,
      this.mainAxisSize,
      this.crossAxisAlignment,
      this.verticalDirection,
      required this.onRefreshEnd});

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ###  Add scroll controller  to can control the scroll widget
  ///
  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// ### اضف
  /// ### scroll controller
  ///### لكي تستطيع التحكم بالويدجت

  final ScrollController controller;

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ###  Add Here Widgets you want to show in the screen
  ///
  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// ### هنا اضف المكونات التي تريدها بداخل الويدجت من نوع
  /// ### Widget

  final List<Widget> children;

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ###  Here the widget shown when the user pulling down
  /// ###  it's  CircularProgressIndicator() by default
  ///
  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// ### هنا اختار الويدجت التي تريدها تظهر للمستخدم عندما يسحب للأسفل
  /// ### CircularProgressIndicator() بشكل افتراضي
  final Widget? loadingWidget;

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ###  Color of the arrow icons it's gray by default
  ///
  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// ### هنا لون ايقونات الأسهم ، لونها رصاصي بشكل افتراضي

  final Color? iconsColor;

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ###  Add here the functions you want when the user pull down
  ///
  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// ### هنا اضف الأوامر التي تريدها عندما المستخدم يسحب للأسفل

  final void Function() onRefresh;

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ###  Add here the functions you want when the refreshing end
  ///
  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// ### هنا اضف الأوامر التي تريدها عندما ينتهي التحديث
  final void Function() onRefreshEnd;

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ###  Here you can change the time taken to refresh before hiding the refresh widget
  /// ###  it's 3 sec by default
  ///
  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// ### من هنا بإمكانك تغيير الوقت المستغرق في التحديث قبل لا يقوم بإخفاء ويدجت التحديث
  /// ### الان يأخذ ٣ ثواني بشكل افتراضي
  final Duration? refreshTime;

  final MainAxisAlignment? mainAxisAlignment;
  final MainAxisSize? mainAxisSize;
  final CrossAxisAlignment? crossAxisAlignment;
  final TextDirection? textDirection;
  final VerticalDirection? verticalDirection;
  final TextBaseline? textBaseline;
  @override
  State<RefreshScrollColumn> createState() => _RefreshScrollColumnState();
}

class _RefreshScrollColumnState extends State<RefreshScrollColumn> {
  int height = 0;
  Widget refreshWidget = const SizedBox.shrink();
  bool isLoading = false;
  bool isDone = false;
  @override
  void initState() {
    Widget loadingWidget =
        widget.loadingWidget ?? const CircularProgressIndicator();
    Duration refreshTime = widget.refreshTime ?? const Duration(seconds: 3);
    widget.controller.addListener(() async {
      if (widget.controller.position.minScrollExtent >
          widget.controller.offset) {
        if (widget.controller.offset <= -150) {
          setState(() {
            height = 200;
            refreshWidget = const Icon(
              Icons.arrow_downward,
              color: Colors.grey,
            );
          });
        } else if (widget.controller.offset <= -70 && height == 200) {
          setState(() {
            height = 100;
            refreshWidget = const Icon(
              Icons.arrow_upward,
              color: Colors.grey,
            );
          });
        } else if (widget.controller.offset <= -50 && height == 100) {
          setState(() {
            height = 50;
            refreshWidget = loadingWidget;
          });

          if (isLoading == false) {
            setState(() {
              widget.onRefresh();
              isDone = true;
            });
          }
          setState(() {
            isLoading = true;
          });
          Timer(refreshTime, () {
            refreshWidget = const SizedBox.shrink();
            setState(() {
              height = 0;
              isLoading = false;
            });
            if (isDone == true) {
              setState(() {
                widget.onRefreshEnd();
                isDone = false;
              });
            }
          });
        }
      }
    });
    // TODO: implement initState
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: MediaQuery.of(context).size.width,
      height: MediaQuery.of(context).size.height,
      child: Column(
        mainAxisAlignment: widget.mainAxisAlignment ?? MainAxisAlignment.start,
        mainAxisSize: widget.mainAxisSize ?? MainAxisSize.max,
        crossAxisAlignment:
            widget.crossAxisAlignment ?? CrossAxisAlignment.start,
        textDirection: widget.textDirection,
        textBaseline: widget.textBaseline,
        verticalDirection: widget.verticalDirection ?? VerticalDirection.down,
        children: [
          AnimatedContainer(
            clipBehavior: Clip.none,
            alignment: Alignment.center,
            padding: const EdgeInsets.symmetric(vertical: 0, horizontal: 50),
            curve: Curves.easeIn,
            duration: const Duration(milliseconds: 300),
            // color: Colors.red,
            width: MediaQuery.of(context).size.width,
            height: height.toDouble(),
            child: refreshWidget,
          ),
          Expanded(
            child: SingleChildScrollView(
              physics: const BouncingScrollPhysics(
                  parent: AlwaysScrollableScrollPhysics()),

              controller: widget.controller,

              child: Column(
                mainAxisAlignment:
                    widget.mainAxisAlignment ?? MainAxisAlignment.start,
                mainAxisSize: widget.mainAxisSize ?? MainAxisSize.max,
                crossAxisAlignment:
                    widget.crossAxisAlignment ?? CrossAxisAlignment.start,
                textDirection: widget.textDirection,
                textBaseline: widget.textBaseline,
                verticalDirection:
                    widget.verticalDirection ?? VerticalDirection.down,
                children: widget.children,
              ), // wireframe for each widget.
            ),
          ),
        ],
      ),
    );
  }
}


lib/src/pull_refresh_listview.dart
import 'dart:async';

import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

class RefreshListViewBuilder extends StatefulWidget {
  const RefreshListViewBuilder({
    Key? key,
    required this.controller,
    this.loadingWidget,
    this.iconsColor,
    required this.onRefresh,
    required this.itemBuilder,
    this.primary,
    this.physics,
    this.padding,
    this.itemExtent,
    this.prototypeItem,
    this.findChildIndexCallback,
    this.itemCount,
    this.cacheExtent,
    this.semanticChildCount,
    this.restorationId,
    this.refreshTime,
    this.onRefreshEnd,
    this.mainAxisAlignment,
    this.mainAxisSize,
    this.crossAxisAlignment,
    this.textDirection,
    this.verticalDirection,
    this.textBaseline,
  }) : super(key: key);

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ###  Add scroll controller  to can control the scroll widget
  ///
  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// ### اضف
  /// ### scroll controller
  ///### لكي تستطيع التحكم بالويدجت

  final ScrollController controller;

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ###  Here the widget shown when the user pulling down
  /// ###  it's  CircularProgressIndicator() by default
  ///
  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// ### هنا اختار الويدجت التي تريدها تظهر للمستخدم عندما يسحب للأسفل
  /// ### CircularProgressIndicator() بشكل افتراضي
  final Widget? loadingWidget;

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ###  Color of the arrow icons it's gray by default
  ///
  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// ### هنا لون ايقونات الأسهم ، لونها رصاصي بشكل افتراضي

  final Color? iconsColor;

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ###  Add here the functions you want when the user pull down
  ///
  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// ### هنا اضف الأوامر التي تريدها عندما المستخدم يسحب للأسفل
  final void Function() onRefresh;

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ###  Add here the functions you want when the refreshing end
  ///
  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// ### هنا اضف الأوامر التي تريدها عندما ينتهي التحديث
  final void Function()? onRefreshEnd;

  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- EN ----</p>
  /// ###  Here you can change the time taken to refresh before hiding the refresh widget
  /// ###  it's 3 sec by default
  ///
  /// <p style="color:#77b4ff;font-size:16;font-weight: bold;">---- AR ----</p>
  /// ### من هنا بإمكانك تغيير الوقت المستغرق في التحديث قبل لا يقوم بإخفاء ويدجت التحديث
  /// ### الان يأخذ ٣ ثواني بشكل افتراضي
  final Duration? refreshTime;
  final NullableIndexedWidgetBuilder itemBuilder;

  final Axis scrollDirection = Axis.vertical;
  final bool reverse = false;
  final bool? primary;
  final ScrollPhysics? physics;
  final bool shrinkWrap = false;
  final EdgeInsetsGeometry? padding;
  final double? itemExtent;
  final Widget? prototypeItem;
  final int? Function(Key)? findChildIndexCallback;
  final int? itemCount;
  final bool addAutomaticKeepAlives = true;
  final bool addRepaintBoundaries = true;
  final bool addSemanticIndexes = true;
  final double? cacheExtent;
  final int? semanticChildCount;
  final DragStartBehavior dragStartBehavior = DragStartBehavior.start;
  final ScrollViewKeyboardDismissBehavior keyboardDismissBehavior =
      ScrollViewKeyboardDismissBehavior.manual;
  final String? restorationId;
  final Clip clipBehavior = Clip.hardEdge;
  final MainAxisAlignment? mainAxisAlignment;
  final MainAxisSize? mainAxisSize;
  final CrossAxisAlignment? crossAxisAlignment;
  final TextDirection? textDirection;
  final VerticalDirection? verticalDirection;
  final TextBaseline? textBaseline;
  @override
  State<RefreshListViewBuilder> createState() => _RefreshListViewBuilderState();
}

class _RefreshListViewBuilderState extends State<RefreshListViewBuilder> {
  int height = 0;
  Widget refreshWidget = const SizedBox.shrink();
  bool isLoading = false;
  bool isDone = false;
  @override
  void initState() {
    Widget loadingWidget =
        widget.loadingWidget ?? const CircularProgressIndicator();
    Duration refreshTime = widget.refreshTime ?? const Duration(seconds: 3);

    widget.controller.addListener(() async {
      if (widget.controller.position.minScrollExtent >
          widget.controller.offset) {
        if (widget.controller.offset <= -150) {
          setState(() {
            setState(() {
              height = 200;
              refreshWidget = const Icon(
                Icons.arrow_downward,
                color: Colors.grey,
              );
            });
          });
        } else if (widget.controller.offset <= -70 && height == 200) {
          setState(() {
            height = 100;
            refreshWidget = const Icon(
              Icons.arrow_upward,
              color: Colors.grey,
            );
          });
        } else if (widget.controller.offset <= -50 && height == 100) {
          setState(() {
            height = 50;
            refreshWidget = loadingWidget;
          });

          if (isLoading == false) widget.onRefresh();
          setState(() {
            isLoading = true;
          });
          Timer(refreshTime, () {
            refreshWidget = const SizedBox.shrink();
            setState(() {
              height = 0;
              isLoading = false;
            });
            if (isDone == true) {
              setState(() {
                widget.onRefreshEnd!();
                isDone = false;
              });
            }
          });
        }
      }
    });
    // TODO: implement initState
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      mainAxisAlignment: widget.mainAxisAlignment ?? MainAxisAlignment.start,
      mainAxisSize: widget.mainAxisSize ?? MainAxisSize.max,
      crossAxisAlignment: widget.crossAxisAlignment ?? CrossAxisAlignment.start,
      textDirection: widget.textDirection,
      textBaseline: widget.textBaseline,
      verticalDirection: widget.verticalDirection ?? VerticalDirection.down,
      children: [
        AnimatedContainer(
          clipBehavior: Clip.none,
          alignment: Alignment.center,
          padding: const EdgeInsets.symmetric(vertical: 0, horizontal: 50),
          curve: Curves.easeIn,
          duration: const Duration(milliseconds: 300),
          // color: Colors.red,
          width: MediaQuery.of(context).size.width,
          height: height.toDouble(),
          child: refreshWidget,
        ),
        Expanded(
          child: ListView.builder(
            itemCount: widget.itemCount,
            primary: widget.primary,
            shrinkWrap: widget.shrinkWrap,
            scrollDirection: widget.scrollDirection,

            physics: widget.physics,
            padding: widget.padding,
            itemExtent: widget.itemExtent,
            prototypeItem: widget.prototypeItem,
            findChildIndexCallback: widget.findChildIndexCallback,
            addAutomaticKeepAlives: widget.addAutomaticKeepAlives,
            addRepaintBoundaries: widget.addRepaintBoundaries,
            addSemanticIndexes: widget.addSemanticIndexes,
            cacheExtent: widget.cacheExtent,
            clipBehavior: widget.clipBehavior,
            semanticChildCount: widget.semanticChildCount,
            dragStartBehavior: widget.dragStartBehavior,
            keyboardDismissBehavior: widget.keyboardDismissBehavior,
            restorationId: widget.restorationId,
            controller: widget.controller,

            itemBuilder: widget.itemBuilder,

            // wireframe for each widget.
          ),
        ),
      ],
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:pull_refresh_widget/pull_refresh_widget.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // TRY THIS: Try running your application with "flutter run". You'll see
        // the application has a blue toolbar. Then, without quitting the app,
        // try changing the seedColor in the colorScheme below to Colors.green
        // and then invoke "hot reload" (save your changes or press the "hot
        // reload" button in a Flutter-supported IDE, or press "r" if you used
        // the command line to start the app).
        //
        // Notice that the counter didn't reset back to zero; the application
        // state is not lost during the reload. To reset the state, use hot
        // restart instead.
        //
        // This works for code too, not just values: Most code changes can be
        // tested with just a hot reload.
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  ScrollController _controller = ScrollController();
  String text = "Pull To Refresh";
  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // TRY THIS: Try changing the color here to a specific color (to
        // Colors.amber, perhaps?) and trigger a hot reload to see the AppBar
        // change color while the other colors stay the same.
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Container(
        // color: Colors.red,

        width: MediaQuery.of(context).size.width,
        height: MediaQuery.of(context).size.height,
        child: RefreshScrollColumn(
          crossAxisAlignment: CrossAxisAlignment.center,
          controller: _controller,
          onRefresh: () {
            print('Refreshing');
            setState(() {
              text = 'Refreshing';
            });
          },
          onRefreshEnd: () {
            print('Refreshed');
            setState(() {
              text = 'Refreshed';
            });
          },
          children: [
            Container(
              alignment: Alignment.center,
              height: 100,
              width: 300,
              child: Text(
                text,
                style: TextStyle(color: Colors.red, fontSize: 30),
              ),
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
